### PR TITLE
WalletConnect PoC

### LINKED ISSUE #335: Implement WalletConnect Logic
We want to be able to use WalletConnect. For this it is required that we can connect to a WalletConnect bridge via websockets and then establish a session with a peer.

### PR DESCRIPTION
**Pull request by rmeissner**
Closes #335 .

Changes proposed in this pull request:
- WalletConnect logic
- PoC UI for WalletConnect management

Note:
Currently approving a wallet connect session will approve it for all available Safes, so the dapp will know all Safe addresses of the app.

@gnosis/mobile-devs

### CODE REVIEW

File: app/src/main/java/pm/gnosis/heimdall/data/repositories/impls/WalletConnectBridgeRepository.kt [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+    }
+        .subscribeOn(Schedulers.io())
+
+    private fun startBridgeService() {
+        context.startService(Intent(context, BridgeService::class.java))
+    }
+}
+
+
+interface Session {
```
Comment (rmeissner) [inline app/src/main/java/pm/gnosis/heimdall/data/repositories/impls/WalletConnectBridgeRepository.kt]: This and other walletconnect logic releated classes should be moved to a separate sdk

File: app/src/main/java/pm/gnosis/heimdall/data/repositories/impls/wc/FileWCSessionStore.kt [hunk 2]
```diff
@@ -0,0 +1,46 @@
+package pm.gnosis.heimdall.data.repositories.impls.wc
+
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.Types
+import pm.gnosis.utils.nullOnThrow
+import java.io.File
+import java.util.concurrent.ConcurrentHashMap
+
+class FileWCSessionStore(private val storageFile: File, moshi: Moshi) : WCSessionStore {
```
Comment (rmeissner) [inline app/src/main/java/pm/gnosis/heimdall/data/repositories/impls/wc/FileWCSessionStore.kt]: We should implement a room based version of this

File: app/src/main/java/pm/gnosis/heimdall/data/repositories/impls/wc/MoshiPayloadAdapter.kt [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+    private fun jsonRpc(id: Long, method: String, vararg params: Any) =
+        mapOf<String, Any>(
+            "id" to id,
+            "jsonrpc" to "2.0",
+            "method" to method,
+            "params" to params
+        )
+
+    @JsonClass(generateAdapter = true)
+    data class EncryptedPayload(val data: String, val iv: String, val hmac: String)
```
Comment (rmeissner) [inline app/src/main/java/pm/gnosis/heimdall/data/repositories/impls/wc/MoshiPayloadAdapter.kt]: annotate parameter names

File: app/src/main/java/pm/gnosis/heimdall/ui/walletconnect/WalletConnectSessionsAdapter.kt [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+                    .switchMapCompletable { viewModel.approveSession(meta.id) }
+                    .subscribeBy(onError = Timber::e)
+
+                disposables += itemView.layout_wallet_connect_sessions_item_deny.clicks()
+                    .switchMapCompletable { viewModel.denySession(meta.id) }
+                    .subscribeBy(onError = Timber::e)
+            }
+        }
+
+        private fun handleSessionEven(event: BridgeRepository.SessionEvent) {
```
Comment (fmrsabino) [inline app/src/main/java/pm/gnosis/heimdall/ui/walletconnect/WalletConnectSessionsAdapter.kt]: handleSessionEven**t**

File: app/src/main/java/pm/gnosis/heimdall/data/repositories/impls/wc/FileWCSessionStore.kt [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+            String::class.java,
+            WCSessionStore.State::class.java
+        )
+    )
+
+    private val currentStates: MutableMap<String, WCSessionStore.State> =
+        ConcurrentHashMap()
+
+    init {
+        val storeContent = storageFile.readText()
```
[the remark this answers is not in the document]
Comment (rmeissner) [reply app/src/main/java/pm/gnosis/heimdall/data/repositories/impls/wc/FileWCSessionStore.kt]: yes, this needs to be refactored into a proper store later anyways

File: app/src/main/java/pm/gnosis/heimdall/data/repositories/impls/wc/OkHttpTransport.kt [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+
+import com.squareup.moshi.Moshi
+import com.squareup.moshi.Types
+import okhttp3.*
+import pm.gnosis.heimdall.data.repositories.impls.Session
+import timber.log.Timber
+import java.util.*
+import java.util.concurrent.ConcurrentLinkedQueue
+
+class OkHttpTransport(
```
[the remark this answers is not in the document]
Comment (rmeissner) [reply app/src/main/java/pm/gnosis/heimdall/data/repositories/impls/wc/OkHttpTransport.kt]: OkHttp uses background threads for this.

[the remark this answers is not in the document]
Comment (rmeissner) [reply app/src/main/java/pm/gnosis/heimdall/data/repositories/impls/wc/OkHttpTransport.kt]: see OkHttpClient source and the Dispatcher class

**Comment by rmeissner** [comment]
Added separate ticket for test coverage https://github.com/gnosis/safe/issues/139

**Pull request merged**